Created new fake user type 'TestStudent'

ANUSHIYAPRIYA opened the pull request:
<!--- Provide a summary of your changes in the Pull Request Title above. -->
<!--- If this is a work in progress (not yet ready to be merged), make this a draft pull request. -->

## Motivation and Context
<!--- Why is this pull request required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
This is the first step for the project 'Feature: Autotest Solution Table'.
The instructor can submit solutions for their tests when setting up automated testing and they can run tests on their solutions.
https://openproject.teach.cs.toronto.edu/projects/markus/work_packages/50/activity



## Your Changes
<!--- Describe your changes here. -->
<!--- Include how your changes may affect other areas of the application, if relevant. -->
**Description**:
In order to treat the solution file as same as the submission file, I have created a fake user type 'Test Student'.
So that the instructor can test their solutions with the submitted test files.


**Type of change** (select all that apply):
<!--- Put an `x` in all the boxes that apply. -->
<!--- Remove any lines that do not apply. --> 


- [x] Breaking change (fix or feature that would cause existing functionality to change)



## Testing
<!--- Please describe in detail how you tested this pull request. -->
<!--- This can include tests you added and manual testing through the web interface. -->
Added tests for the changes

## Questions and Comments (if applicable)
<!-- Ask any questions you have for the maintainers of this project regarding this PR. -->
<!-- Please describe the steps you have already taken to find the answer to your question. -->
<!-- This will ensure that we can give you clear and relevant advice. -->
<!-- If you have additional comments add them here as well. -->


## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have fixed any Hound bot comments. <!-- (check after opening pull request) -->
- [x] I have verified that the TravisCI tests have passed. <!-- (check after opening pull request) -->
- [x] I have reviewed the test coverage changes reported on Coveralls. <!-- (check after opening pull request) -->
- [x] I have added tests for my changes. <!-- (delete this checklist item if not applicable) -->
- [ ] I have updated the Changelog.md file. <!-- (delete this checklist item if not applicable) -->


### Required documentation changes (if applicable)

mishaschwartz reviewed (changes requested):
Some other things that should be included in this PR:

- Make sure that a grouping with a TestStudent is a member is always a valid grouping
- Make sure that the TestStudent's groupings are never shown in the UI
- Make/change any policies that are relevant to the new user
- Add relevant associations to the TestStudent model
- Add rspec tests for the TestStudent model
- etc.

ANUSHIYAPRIYA commented:
> Some other things that should be included in this PR:
> 
> * Make sure that a grouping with a TestStudent is a member is always a valid grouping
> * Make sure that the TestStudent's groupings are never shown in the UI
> * Make/change any policies that are relevant to the new user
> * Add relevant associations to the TestStudent model
> * Add rspec tests for the TestStudent model
> * etc.

Points which are addressed:
1. Validating associated grouping. After adding the code for creating a grouping for test student, will add more validation
2. Made sure that the TestStudent's groupings are never shown in the UI
3. Added relevant associations to the TestStudent model
4. Added rspec tests for the TestStudent model

Points which are not necessary:
1. Make/change any policies that are relevant to the new user (I think for the upcoming PR's we may need some changes in policy.)

mishaschwartz commented on app/models/grouping.rb:
```diff
@@ -25,6 +25,8 @@ class Grouping < ApplicationRecord
   has_many :ta_memberships, class_name: 'TaMembership'
   has_many :tas, through: :ta_memberships, source: :user
   has_many :students, through: :student_memberships, source: :user
+  has_many :test_students, class_name: 'TestStudent', through: :student_memberships, source: :user
```
I'm not sure if this is necessary. I don't think the grouping object needs to differentiate between students and test_students here.

mishaschwartz commented on app/models/test_student.rb:
```diff
@@ -0,0 +1,14 @@
+# TestStudent - fake student, used for running tests on solution files
+class TestStudent < User
+  has_many :accepted_groupings, lambda {
```
use the arrow syntax `->` instead of `lambda`

ANUSHIYAPRIYA replied:
I used the -> first but Rubocop raised an offense and suggested using lambda. Which one is recommended?

mishaschwartz replied:
I'm going to make a decision here and say that we should go with `->`. I'll make a separate PR that disables that cop so houndbot won't complain in the future.

mishaschwartz reviewed (changes requested):
Some additional validations and checks I would like to see added:
- a grouping with a test student as a member should have no other members
- a test student should not be allowed to be invited to join a grouping
- a test student can only be an 'inviter'

are there any others you think are appropriate to add?

ANUSHIYAPRIYA commented:
> Some additional validations and checks I would like to see added:
> 
> * a grouping with a test student as a member should have no other members
> * a test student should not be allowed to be invited to join a grouping
> * a test student can only be an 'inviter'
> 
> are there any others you think are appropriate to add?

Points addressed:
1. A grouping with a test student as a member should have no other members
2. A test student can only be an 'inviter'

Points not addressed:
1. A test student should not be allowed to be invited to join a grouping
Reason:
By default, the existing code does this behavior. It is only allowing the students of type 'Student' and the unhidden students. I checked all the corresponding method for inviting a member. Those are all already checking for the class as 'Student' class and hidden is false. So this will automatically exclude the test students.

mishaschwartz commented on app/models/grouping.rb:
```diff
@@ -778,4 +779,11 @@ def use_section_due_date?
   end
+
+  def test_student_grouping_member
+    members = memberships.joins(:user).where(grouping_id: self.id).count
```
We care about `accepted_student_memberships` here not all `memberships`

Try starting with:

```
accepted_student_memberships.joins(:user).pluck('users.type')
```

and see if you can find a more efficient way of checking this

mishaschwartz commented on app/models/grouping.rb:
```diff
@@ -83,7 +84,7 @@ class Grouping < ApplicationRecord
   has_many :grouping_starter_file_entries, dependent: :destroy
   has_many :starter_file_entries, through: :grouping_starter_file_entries
-
+  validate :test_student_grouping_member, if: -> { !self.id.nil? && self.test_students.exists? }
```
use `new_record?` instead of checking for nil.

also, can we do this validation with having a `test_students` association please.

also, this validation should be doable even if we have a new record

mishaschwartz commented on app/models/student_membership.rb:
```diff
@@ -82,4 +84,14 @@ def one_accepted_per_assignment
   end
+
+  def test_student_membership_status
+    user.memberships.each do |membership|
```
this validation should only care about the membership referred to by `self`, not all memberships belonging to the user.

try something like:

```ruby
errors.add(...) if user.test_student? && membership_status != STATUSES[:inviter]
```

mishaschwartz commented on app/models/student_membership.rb:
```diff
@@ -82,4 +84,14 @@ def one_accepted_per_assignment
+    user.memberships.each do |membership|
+      unless membership.membership_status == 'inviter'
+        errors.add(:base, 'A test student can only be an inviter')
+        false
```
validations don't need to return true or false

mishaschwartz commented on spec/factories/test_students.rb:
```diff
@@ -0,0 +1,5 @@
+FactoryBot.define do
+  factory :test_student, class: TestStudent, parent: :user do
+    hidden { true }
```
this is no longer necessary

mishaschwartz commented on app/models/student_membership.rb:
```diff
@@ -82,4 +84,14 @@ def one_accepted_per_assignment
+
+  def test_student_membership_status
+    user.memberships.each do |membership|
+      unless membership.membership_status == 'inviter'
```
always refer to the `STATUSES` hash, not the string value directly

mishaschwartz commented on app/models/grouping.rb:
```diff
@@ -83,7 +84,8 @@ class Grouping < ApplicationRecord
   has_many :starter_file_entries, through: :grouping_starter_file_entries
-
+  validate :test_student_grouping_member, if: -> { self.test_students.exists? }
+  validates_associated :test_students
```
this can be removed
